Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,191 @@
+        }
+
+        return messages
+    }
+
+    // MARK: Private
+
+    /// Seeks the reader and writer file handles to the correct place.
+    /// Only necessary the first time this method is called.
+    private func setUpFileHandlesIfNecessary() throws {
+        guard !hasSetUpFileHandles else {
+            return
+        }
+
+        // This is our first cache action.
+        // We need to find out where we should write our next message.
+        try reader.seekToEndOfNewestMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages)
+        try writer.seek(toOffset: reader.offsetInFile)
+
+        // Now that we know where to write, we need to figure out where the oldest message is.
+        try reader.seekToBeginningOfOldestMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages)
+
+        hasSetUpFileHandles = true
+    }
+
+    /// Writes message data to the cache.
+    ///
+    /// For caches that do not roll, the message data is written in the following format:
+    /// `[messageData][endOfNewestMessageMarker]`
+    /// - `messageData` is an `EncodableMessage`'s encoded data.
+    /// - `endOfNewestMessageMarker` is a big-endian encoded `MessageSpan` of length `messageSpanLength`.
+    ///
+    /// For caches that roll, the message data is written in the following format:
+    /// `[messageData][endOfNewestMessageMarker][offsetInFileOfOldestMessage]`
+    /// - `messageData` is an `EncodableMessage`'s encoded data.
+    /// - `endOfNewestMessageMarker` is a big-endian encoded `MessageSpan` of length `messageSpanLength`.
+    /// - `offsetInFileOfOldestMessage` is a big-endian encoded `Bytes` of length `oldestMessageOffsetLength`.
+    ///
+    /// By the time this method returns, the `writer`'s `offsetInFile` is always set to the
+    /// beginning of the written `endOfNewestMessageMarker`, such that the next message
+    /// written will overwrite the marker.
+    ///
+    /// - Parameter messageData: an `EncodableMessage`'s encoded data. Must be smaller than both `maximumBytes` and `MessageSpan.max`.
+    private func write(messageData: Data) throws {
+        let messageLength = Bytes(messageData.count) + lengthOfMessageSuffix
+        guard messageLength <= maximumBytes && messageLength <= Bytes(MessageSpan.max) else {
+            throw CacheAdvanceWriteError.messageDataTooLarge
+        }
+
+        while shouldOverwriteOldMessages
+            && writer.offsetInFile < reader.offsetInFile
+            && reader.offsetInFile < writer.offsetInFile + messageLength
+        {
+            // We are a rolling cache. The current position of the writer is before the oldest message,
+            // and writing this message would write into the current message. Advance to the next message.
+            try reader.seekToNextMessage(shouldSeekToOldestMessageIfFound: true, cacheOverwritesOldMessages: shouldOverwriteOldMessages)
+        }
+
+        try writer.__write(messageData, error: ())
```

Comment:
What's going on with passing `()` to `error:`?

Reply:
Also where do these underscores come from?

Reply:
Good questions! These all come from Apple's API:
<img width="792" alt="image" src="https://user-images.githubusercontent.com/139364/69000690-995ffa80-0888-11ea-86e5-0d26a2185d03.png">

It looks like the original, non-`__` versions all threw objective-c exceptions, not Swift errors, and are now unofficially deprecated:
<img width="771" alt="image" src="https://user-images.githubusercontent.com/139364/69000697-b3014200-0888-11ea-9d36-0bc6b0d29a91.png">

It is indeed super-weird that `__write` takes a `Void` error... but I'm just following what the API declares.